RRD structures managed by dictionaries

ktsaou opened the pull request:
This PR moves RRD structures so that they are fully managed by dictionaries.

Once this finishes, we are going to have reference counters, so that each RRD object will remain in the dictionary (and allocated) for as long as there is a reference to it. We expect that this will greatly modernize our code base, simplify lock management and increase the reliability and stability of the agent.

## dictionaries improvements

- [x] extent dictionaries to allow optimal construction for heavy objects (delay the allocation of the members until is really needed, if needed - it may not in case of a conflict). This has been done by providing a `dictionary_set_advanced()` that accepts an optional constructor object with the values to be used for initializing the target object. It is passed to the `data` parameter of the `insert`, `conflict` and `react` callbacks.
- [x] dictionaries registered callbacks should get an acquired dictionary item, instead of the name
- [x] dictionaries walkthrough callbacks should get an acquired dictionary item, instead of the name
- [ ] dictionaries callbacks once they get the dictionary item, can be used with 2 parameters, not 3. The value can be found via `dictionary_acquired_item_value()`.
- [x] allow duping a zero-referenced dictionary item - review dictionaries reference counter and provide a unit test to ensure they work as expected
- [x] find a way so that secondary indexes can be atomically operated and that acquiring a object from a secondary index acquires the primary object.
- [x] remove all the `unsafe` versions of dictionaries - try to find a way for dictionaries to know that the caller already has write access to the dictionary.

## Manage memory of objects via dictionaries
So that we can ensure they will survive while they are acquired by another thread. This turns them into somewhat object-oriented for the construction, the update (conflict) and the destruction.

- [ ] rrdhost
- [x] rrdset
- [x] rrddim
- [x] rrdsetvar
- [x] rrddimvar
- [x] rrdfamily
- [x] rrdcalc
- [x] rrdcalctemplate
- [x] rrdvar

## Eliminate linked lists
Replace linked lists of objects with dictionary traversals (dfe or walkthrough), so that the locks of the objects are no longer protecting the linked list. The dictionaries protect themselves.

- [ ] rrdhost
- [x] rrdset
- [x] rrddim (done, but the query engine requires a new solution to allow both chart and context queries - see below)
- [x] rrdsetvar
- [x] rrddimvar
- [x] rrdcalc (the linked list survived as an optimization to allow /api/v1/charts go faster)
- [x] rrdcalctemplate
- [ ] review all dictionary walkthroughs and make sure long standing rrdhost and rrdset walkthroughs are configured as `reentrant`

## Locks
Review all remaining `RRDHOST` and `RRDSET` locks to figure out the reason they exist.

- [ ] provide new lock to control who can save/cleanup the database at any given time (for db mode "save" and "map")
- [ ] more to come here...

## Acquire/Release for RRD objects
RRD objects should be completely hidden from each other and when references are required, they should be via acquire/release.

- [ ] rrdhost
- [ ] rrdset (too complex for the moment, postponed for later)
- [ ] rrddim (too complex for the moment, postponed for later)
- [x] rrdsetvar
- [x] rrddimvar (this has been completely hidden - no one gets any output from this - they are exposed as rrdvar)
- [x] rrdvar (done, although proc plugin never releases them)
- [x] rrdfamily
- [x] rrdcalc (was needed for budges)
- [x] rrdcalctemplate (none was needed)
- [ ] find a way for internal plugins to release their acquired items during cleanup (currently needed for host and chart variables in proc.plugin and cgroups.plugin)

## Query engine based on rrdcontexts

- [ ] make rrdcontexts permanent (should never be disabled by users)
- [ ] use rrdcontexts to drive the query engine, eliminating the need for `RRDSET` and `RRDDIM` for context queries. This is required to get rid of the RRDDIM linked list.

thiagoftsm commented on database/sqlite/sqlite_functions.c:
```diff
@@ -697,20 +697,21 @@ void delete_dimension_uuid(uuid_t *dimension_uuid)
+int find_chart_uuid(RRDHOST *host, const char *type, const char *id, const char *name, uuid_t *store_uuid)
 {
     static __thread sqlite3_stmt *res = NULL;
     uuid_t *uuid = NULL;
```
I am starting the review tonight, and I observed when I compiled this branch I am receiving this warning:

```sh
database/sqlite/sqlite_functions.c: In function 'find_chart_uuid':
database/sqlite/sqlite_functions.c:703:13: warning: unused variable 'uuid' [-Wunused-variable]
  703 |     uuid_t *uuid = NULL;
      | 
```

ktsaou replied:
@stelfrag this is yours...

thiagoftsm commented:
Hello @ktsaou,

When I started this PR with `[db].mode = map` netdata did not start and I got this [error.log](https://github.com/netdata/netdata/files/9539280/error.log). These logs were not showing useful information so I used  `gdb --args /usr/sbin/netdata -D` and I got this:

```sh
Reading symbols from /usr/sbin/netdata...
(gdb) r
Starting program: /usr/sbin/netdata -D
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
[New Thread 0x7ffff29fd6c0 (LWP 20954)]
[Detaching after fork from child process 20955]
[Detaching after vfork from child process 20956]
[New Thread 0x7ffff20f96c0 (LWP 21157)]
[New Thread 0x7ffff18f86c0 (LWP 21158)]

Thread 1 "netdata" received signal SIGSEGV, Segmentation fault.
0x0000000000c3e5d1 in rrdsetvar_create_variables (rs=rs@entry=0x60b000033680) at database/rrdsetvar.c:42
42          RRDHOST *host = st->rrdhost;
(gdb) bt
#0  0x0000000000c3e5d1 in rrdsetvar_create_variables (rs=rs@entry=0x60b000033680) at database/rrdsetvar.c:42
#1  0x0000000000c3fce3 in rrdsetvar_insert_callback (name=<optimized out>, rrdsetvar=0x60b000033680, constructor_data=0x7fffffff5a60) at database/rrdsetvar.c:124
#2  0x0000000000560ada in item_create_unsafe (constructor_data=<optimized out>, value_len=<optimized out>, value=<optimized out>, name_len=<optimized out>, name=<optimized out>, dict=0x61200016aec0)
    at libnetdata/dictionary/dictionary.c:679
#3  dictionary_set_item_unsafe (dict=dict@entry=0x61200016aec0, name=name@entry=0x16525c0 "last_collected_t", name_len=<optimized out>, name_len@entry=17, value=value@entry=0x0, value_len=value_len@entry=112, 
    constructor_data=constructor_data@entry=0x7fffffff5a60) at libnetdata/dictionary/dictionary.c:916
#4  0x000000000056501e in dictionary_set_advanced (dict=0x61200016aec0, name=name@entry=0x16525c0 "last_collected_t", name_len=17, name_len@entry=-1, value=value@entry=0x0, value_len=value_len@entry=112, 
    constructor_data=constructor_data@entry=0x7fffffff5a60) at libnetdata/dictionary/dictionary.c:1001
#5  0x0000000000c401d2 in rrdsetvar_create (st=st@entry=0x61500009c480, variable=variable@entry=0x16525c0 "last_collected_t", type=type@entry=RRDVAR_TYPE_TIME_T, value=value@entry=0x61500009c5c0, 
    options=options@entry=RRDVAR_OPTION_DEFAULT) at database/rrdsetvar.c:194
#6  0x0000000000c2d2b1 in rrdset_react_callback (chart_full_id=<optimized out>, rrdset=0x61500009c480, constructor_data=0x7fffffff5c40) at database/rrdset.c:350
#7  0x000000000056524a in dictionary_set_advanced (dict=0x6120000091c0, name=name@entry=0x7fffffff5ce0 "anomaly_detection.anomaly_rates", name_len=<optimized out>, name_len@entry=-1, value=value@entry=0x0, 
    value_len=value_len@entry=464, constructor_data=constructor_data@entry=0x7fffffff5c40) at libnetdata/dictionary/dictionary.c:1012
#8  0x0000000000c32c80 in rrdset_index_add (st_ctr=0x7fffffff5c40, id=0x7fffffff5ce0 "anomaly_detection.anomaly_rates", host=0x618000016c80) at database/rrdset.c:408
#9  rrdset_create_custom (host=0x618000016c80, type=<optimized out>, id=0x162c600 "anomaly_rates", name=<optimized out>, family=<optimized out>, context=<optimized out>, title=<optimized out>, units=<optimized out>, 
    plugin=<optimized out>, module=<optimized out>, priority=<optimized out>, update_every=<optimized out>, chart_type=<optimized out>, memory_mode=<optimized out>, history_entries=<optimized out>) at database/rrdset.c:937
#10 0x0000000000b8055a in ml::RrdHost::RrdHost (RH=0x618000016c80, this=0x61600000ff80) at ml/Host.h:18
#11 ml::TrainableHost::TrainableHost (RH=0x618000016c80, this=0x61600000ff80) at ml/Host.h:67
#12 ml::DetectableHost::DetectableHost (this=0x61600000ff80, RH=0x618000016c80) at ml/Host.h:91
#13 0x0000000000b6c941 in ml_new_host (RH=RH@entry=0x618000016c80) at ml/ml.cc:52
#14 0x0000000000c226a1 in rrdhost_create (hostname=hostname@entry=0x602000000910 "hades", registry_hostname=<optimized out>, guid=guid@entry=0x2c7cd00 <guid> "e83bb26c-e205-11ec-8f32-f82819901f7f", os=os@entry=0x1504760 "linux", 
    timezone=timezone@entry=0x602000001430 "UTC", abbrev_timezone=abbrev_timezone@entry=0x602000001450 "UTC", utc_offset=<optimized out>, tags=<optimized out>, program_name=<optimized out>, program_version=<optimized out>, 
    update_every=<optimized out>, entries=<optimized out>, memory_mode=<optimized out>, health_enabled=<optimized out>, rrdpush_enabled=<optimized out>, rrdpush_destination=<optimized out>, rrdpush_api_key=<optimized out>, 
    rrdpush_send_charts_matching=<optimized out>, system_info=<optimized out>, is_localhost=<optimized out>, archived=<optimized out>) at database/rrdhost.c:529
#15 0x0000000000c239a5 in rrd_init (hostname=<optimized out>, system_info=system_info@entry=0x611000007e80) at database/rrdhost.c:931
#16 0x0000000000528a93 in main (argc=<optimized out>, argv=<optimized out>) at daemon/main.c:1465
```

thiagoftsm commented:
After to test on `Slackware current`, I decided to test on `Ubuntu 22.04` and I got analogous result, but this time I was using `systemd`

```sh
Core was generated by `/usr/sbin/netdata -P /run/netdata/netdata.pid -D'.
--Type <RET> for more, q to quit, c to continue without paging--
Program terminated with signal SIGSEGV, Segmentation fault.
#0  rrdsetvar_create_variables (rs=rs@entry=0x55b0489a1310) at database/rrdsetvar.c:42
42          RRDHOST *host = st->rrdhost;
[Current thread is 1 (Thread 0x7fd923162640 (LWP 12052))]
(gdb) bt
#0  rrdsetvar_create_variables (rs=rs@entry=0x55b0489a1310) at database/rrdsetvar.c:42
#1  0x000055b0448ab33c in rrdsetvar_insert_callback (name=<optimized out>, rrdsetvar=0x55b0489a1310, constructor_data=0x7fd923161410) at database/rrdsetvar.c:124
#2  0x000055b044835cb0 in item_create_unsafe (constructor_data=<optimized out>, value_len=<optimized out>, value=<optimized out>, name_len=<optimized out>, name=0xb1 <error: Cannot access memory at address 0xb1>, dict=0x55b0489ab580)
    at libnetdata/dictionary/dictionary.c:679
#3  dictionary_set_item_unsafe (dict=dict@entry=0x55b0489ab580, name=name@entry=0x55b044cc9ddc "last_collected_t", name_len=<optimized out>, name_len@entry=17, value=value@entry=0x0, value_len=value_len@entry=112, 
    constructor_data=constructor_data@entry=0x7fd923161410) at libnetdata/dictionary/dictionary.c:916
#4  0x000055b04483731a in dictionary_set_advanced (dict=0x55b0489ab580, name=name@entry=0x55b044cc9ddc "last_collected_t", name_len=17, name_len@entry=-1, value=value@entry=0x0, value_len=value_len@entry=112, 
    constructor_data=constructor_data@entry=0x7fd923161410) at libnetdata/dictionary/dictionary.c:1001
#5  0x000055b0448ab4e3 in rrdsetvar_create (st=st@entry=0x55b0489a0c60, variable=variable@entry=0x55b044cc9ddc "last_collected_t", type=type@entry=RRDVAR_TYPE_TIME_T, value=value@entry=0x55b0489a0da0, 
    options=options@entry=RRDVAR_OPTION_DEFAULT) at database/rrdsetvar.c:194
#6  0x000055b0448a4986 in rrdset_react_callback (chart_full_id=<optimized out>, rrdset=0x55b0489a0c60, constructor_data=0x7fd923161540) at database/rrdset.c:350
#7  0x000055b044837404 in dictionary_set_advanced (dict=0x55b0484bd560, name=name@entry=0x7fd9231615c0 "netdata.statsd_metrics", name_len=<optimized out>, name_len@entry=-1, value=value@entry=0x0, value_len=value_len@entry=464, 
    constructor_data=constructor_data@entry=0x7fd923161540) at libnetdata/dictionary/dictionary.c:1012
#8  0x000055b0448a69a2 in rrdset_index_add (st_ctr=0x7fd923161540, id=0x7fd9231615c0 "netdata.statsd_metrics", host=0x55b048471460) at database/rrdset.c:408
#9  rrdset_create_custom (host=0x55b048471460, type=type@entry=0x55b044ca6ac2 "netdata", id=id@entry=0x55b044cee1e8 "statsd_metrics", name=name@entry=0x0, family=family@entry=0x55b044ca7d36 "statsd", context=context@entry=0x0, 
    title=0x55b044cef7c0 "Metrics in the netdata statsd database", units=0x55b044d076d8 "metrics", plugin=0x55b044cedefc "statsd.plugin", module=0x55b044d0089f "stats", priority=132010, update_every=1, chart_type=RRDSET_TYPE_STACKED, 
    memory_mode=RRD_MEMORY_MODE_MAP, history_entries=3996) at database/rrdset.c:937
#10 0x000055b0449888a0 in statsd_main (ptr=0x55b04839ea60) at collectors/statsd.plugin/statsd.c:2526
#11 0x000055b04485073f in thread_start (ptr=<optimized out>) at libnetdata/threads/threads.c:187
#12 0x00007fd927ad5b43 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#13 0x00007fd927b67a00 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81

```

ktsaou commented:
Thanks @thiagoftsm !
Solved the problem.

thiagoftsm commented:
Hello @ktsaou ,

The issue was really fixed, but now I detected another, please take a look in this [error.log](https://github.com/netdata/netdata/files/9540260/error.log), this is happening when I am using `[db].memory = ram`, I also ran Netdata in `gdb`:

```sh
Thread 32 "PLUGIN[proc net" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffe2cbc6c0 (LWP 23327)]
0x000000000042e8f5 in __asan::asan_free(void*, __sanitizer::BufferedStackTrace*, __asan::AllocType) ()
(gdb) bt
#0  0x000000000042e8f5 in __asan::asan_free(void*, __sanitizer::BufferedStackTrace*, __asan::AllocType) ()
#1  0x00000000004bd67f in __interceptor_free.part.0 ()
#2  0x0000000000ca7df5 in sqlite3MemFree (pPrior=<optimized out>) at database/sqlite/sqlite3.c:25196
#3  0x0000000000cb97ae in sqlite3_free (p=0xc29438 <chart_label_store_to_sql_callback>) at database/sqlite/sqlite3.c:28981
#4  0x0000000000c4c545 in db_execute (
    cmd=0x619000223d80 "INSERT OR REPLACE INTO chart_label (chart_id, source_type, label_key, label_value, date_created) VALUES (u2h('d83fd937-2784-4ae7-a22e-1d878c1b1919'), 536870913,'_collect_plugin','proc.plugin', unixepo"...)
    at database/sqlite/sqlite_functions.c:1540
#5  0x0000000000c2b81c in rrdset_update_rrdlabels (st=0x615000152500, new_rrdlabels=<optimized out>) at database/rrdlabels.c:971
#6  0x000000000100abd1 in do_proc_net_dev (update_every=<optimized out>, dt=dt@entry=0) at collectors/proc.plugin/proc_net_dev.c:1084
#7  0x000000000100ffad in netdev_main (ptr=<optimized out>) at collectors/proc.plugin/proc_net_dev.c:1511
#8  0x00000000005a873f in thread_start (ptr=<optimized out>) at libnetdata/threads/threads.c:187
#9  0x00007ffff6894fea in start_thread () from /lib64/libc.so.6
#10 0x00007ffff691f9ec in clone3 () from /lib64/libc.so.6
(gdb) 
```

On the other hand on `Ubuntu 22.04`  I did not have any issues. During the day I will retest this PR on other environments and continue reviewing the code. :handshake:

thiagoftsm commented:
Hello @ktsaou ,

After last commits I can run this branch using `[db].mode = ram`, thanks for this fix! :)

thiagoftsm commented:
Hello @ktsaou ,

Last commits added these warnings:

```sh

 --- Compile netdata --- 
[/home/thiago/Netdata/tests_netdata]# make -j4 
In file included from ./libnetdata/libnetdata.h:411,
                 from ./daemon/common.h:6,
                 from database/rrd.h:39,
                 from database/rrdsetvar.c:3:
database/rrdsetvar.c: In function 'rrdsetvar_custom_chart_variable_set':
database/rrdsetvar.c:254:15: warning: format '%x' expects argument of type 'unsigned int', but argument 10 has type 'int' [-Wformat=]
  254 |         error("RRDSETVAR: requested to set variable '%s' of chart '%s' on host '%s' to value " NETDATA_DOUBLE_FORMAT
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  255 |             " but the variable is not a custom chart one (it has options 0x%x, value pointer %p). Ignoring request.", string2str(rs->name), rrdset_id(st), rrdhost_hostname(st->rrdhost), value, rs->flags, rs->value);
      |                                                                                                                                                                                                  ~~~~~~~~~
      |                                                                                                                                                                                                    |
      |                                                                                                                                                                                                    int
./libnetdata/log/log.h:97:81: note: in definition of macro 'error'
   97 | #define error(args...)   error_int("ERROR", __FILE__, __FUNCTION__, __LINE__, ##args)
      |                                                                                 ^~~~
In file included from ml/kmeans/dlib/dlib/all/../logger/extra_logger_headers.h:7,
                 from ml/kmeans/dlib/dlib/all/../logger/extra_logger_headers.cpp:6,
                 from ml/kmeans/dlib/dlib/all/source.cpp:39:
ml/kmeans/dlib/dlib/all/../logger/logger_kernel_1.h: In function 'dlib::log_level dlib::logger_config_file_helpers::string_to_log_level(const std::string&)':
ml/kmeans/dlib/dlib/all/../logger/logger_kernel_1.h:36:20: warning: 'char* __builtin_strncpy(char*, const char*, long unsigned int)' output truncated before terminating nul copying 19 bytes from a string of the same length [-Wstringop-t
runcation]
   36 |             strncpy(name,name_,19);
      |                    ^
database/sqlite/sqlite_functions.c: In function 'find_chart_uuid':
database/sqlite/sqlite_functions.c:703:13: warning: unused variable 'uuid' [-Wunused-variable]
  703 |     uuid_t *uuid = NULL;
      |             ^~~~
libnetdata/dictionary/dictionary.c: In function 'check_item_deleted_flag':
libnetdata/dictionary/dictionary.c:2272:46: warning: format '%u' expects argument of type 'unsigned int', but argument 3 has type 'int32_t' {aka 'int'} [-Wformat=]
 2272 |     fprintf(stderr, "NAME_VALUE refcount is %u, expected %u...\t\t\t\t\t", nv->refcount, refcount);
      |                                             ~^                             ~~~~~~~~~~~~
      |                                              |                               |
      |                                              unsigned int                    int32_t {aka int}
      |                                             %u
libnetdata/dictionary/dictionary.c:2272:59: warning: format '%u' expects argument of type 'unsigned int', but argument 4 has type 'int' [-Wformat=]
 2272 |     fprintf(stderr, "NAME_VALUE refcount is %u, expected %u...\t\t\t\t\t", nv->refcount, refcount);
      |                                                          ~^                              ~~~~~~~~
      |                                                           |                              |
      |                                                           unsigned int                   int
      |                                                          %u
 OK  ''

```

Tests that I ran yesterday did not show any new issue, I will describe the tests later today with a comment.

thiagoftsm commented:
Hello @ktsaou ,

I was preparing a new environment to test streaming and I got this error on parent:

```sh
2022-09-11 18:51:24: netdata FATAL : PLUGIN[proc] : (0978@database/rrdhost.c  :__rrdhost_check): RRDHOST 'hades' should be read-locked, but it is not, at function rrdcalc_add_to_host() at line 399 of file 'database/rrdcalc.c' # : No suc
h file or directory

2022-09-11 18:51:24: netdata LOG FLOOD PROTECTION reset for process 'netdata' (prevented 532 logs in the last 1 seconds).
2022-09-11 18:51:24: netdata INFO  : PLUGIN[proc] : (0031@daemon/main.c       :netdata_cleanup): EXIT: netdata prepares to exit with code 1...
2022-09-11 18:51:24: netdata INFO  : PLUGIN[proc] : (0042@daemon/main.c       :netdata_cleanup): EXIT: cleaning up the database...
2022-09-11 18:51:24: netdata INFO  : PLUGIN[proc] : (1477@database/rrdhost.c  :rrdhost_cleanup): Cleaning up database [3 hosts(s)]...
2022-09-11 18:51:24: netdata INFO  : PLUGIN[proc] : (1429@database/rrdhost.c  :rrdhost_cleanup): Cleaning up database of host 'hades'...
2022-09-11 18:51:24: netdata ERROR : PLUGIN[proc] : (0236@libnetdata/locks/loc:__netdata_rwloc): RW_LOCK: failed to obtain write lock (code 35) (errno 2, No such file or directory)
2022-09-11 18:51:24: netdata ERROR : PLUGIN[proc] : (0223@libnetdata/locks/loc:__netdata_rwloc): RW_LOCK: failed to obtain read lock (code 35)
2022-09-11 18:51:24: netdata INFO  : PLUGIN[proc] : (1429@database/rrdhost.c  :rrdhost_cleanup): Cleaning up database of host 'potinho'...
2022-09-11 18:51:24: netdata INFO  : PLUGIN[proc] : (1429@database/rrdhost.c  :rrdhost_cleanup): Cleaning up database of host 'ubuntu1804.localdomain'...
2022-09-11 18:51:24: netdata INFO  : PLUGIN[proc] : (0115@database/sqlite/sqli:sql_close_conte): Closing context SQLite database
2022-09-11 18:51:24: netdata INFO  : PLUGIN[proc] : (0565@database/sqlite/sqli:sql_close_datab): Closing SQLite database
2022-09-11 18:51:24: netdata INFO  : PLUGIN[proc] : (0145@database/sqlite/sqli:add_stmt_to_lis): No statements pending to finalize
2022-09-11 18:51:24: netdata INFO  : PLUGIN[proc] : (0080@daemon/main.c       :netdata_cleanup): EXIT: all done - netdata is now exiting - bye bye...
```

thiagoftsm commented:
Hello @ktsaou ,

Thank you to fix https://github.com/netdata/netdata/pull/13646/#issuecomment-1243021717 now I will proceed with tests.
Only to remember, these warnings https://github.com/netdata/netdata/pull/13646/#issuecomment-1243011634 are still present during compilation time.

Best regards!

thiagoftsm commented on database/rrdcalc.c:
```diff
@@ -499,274 +568,240 @@ inline RRDCALC *rrdcalc_create_from_template_unsafe(RRDHOST *host, RRDCALCTEMPLA
-    }
-    rrdhost_unlock(host);
+static size_t rrdcalc_key(char *dst, size_t dst_len, const char *chart, const char *alert) {
+    return snprintf(dst, dst_len, "%s/%s", chart, alert);
```
Is there a special reason to use `snprintf` instead `snprintfz`?

ktsaou replied:
thanks! fixed it!

thiagoftsm commented:
Hello guys,

I update the [comment](https://github.com/netdata/netdata/pull/13646#issuecomment-1243011634) removing the ML warning, and now I am adding more warnings that were added today:

```sh
In file included from libnetdata/dictionary/../libnetdata.h:411,
                 from libnetdata/dictionary/dictionary.c:13:
libnetdata/dictionary/dictionary.c: In function 'dictionary_destroy':
libnetdata/dictionary/dictionary.c:938:13: warning: format '%ld' expects argument of type 'long int', but argument 9 has type 'size_t' {aka 'long unsigned int'} [-Wformat=]
  938 |             "DICTIONARY: delaying destruction of dictionary created from %s() %zu@%s, because it has %ld referenced items in it (%ld total).",
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
......
  942 |             dictionary_stats_referenced_items(dict),
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |             |
      |             size_t {aka long unsigned int}
libnetdata/dictionary/../log/log.h:89:127: note: in definition of macro 'internal_error'
   89 | #define internal_error(condition, args...) do { if(unlikely(condition)) error_int("IERR", __FILE__, __FUNCTION__, __LINE__, ##args); } while(0)
      |                                                                                                                               ^~~~
libnetdata/dictionary/dictionary.c:938:104: note: format string is defined here
  938 |             "DICTIONARY: delaying destruction of dictionary created from %s() %zu@%s, because it has %ld referenced items in it (%ld total).",
      |                                                                                                      ~~^
      |                                                                                                        |
      |                                                                                                        long int
      |                                                                                                      %ld
```

ktsaou commented:
@thiagoftsm I have fixed all warnings.

thiagoftsm commented:
@stelfrag and @ktsaou, when I began to test streaming this morning, I saw this error on `parent`:

```sh
2022-09-13 11:54:46: netdata IERR  : rrdcontext : (2873@database/rrdcontext.:check_if_cloud_): RRDCONTEXT: SENDING NEW VERSION 'sensors.current', version 1663070087, title 'Current', units 'Ampere', family 'current', chart type 'line', priority 60003, first_time_t 1663070085 (CHANGED), last_time_t 1663070085 (CHANGED), deleted 'false', (queued for 1663070086360 ms, expected 0 ms)
database/sqlite/sqlite3.c:74389:14: runtime error: load of address 0x7fffe027cad0 with insufficient space for an object of type 'struct MemPage *'
0x7fffe027cad0: note: pointer points here
 20 63 00 00  88 95 17 00 10 62 00 00  88 49 09 00 10 62 00 00  00 cb 27 e0 ff 7f 00 00  fc 43 c8 00
              ^ 
EOF found in spawn pipe.
Shutting down spawn server event loop.
Shutting down spawn server loop complete.
```

Logs were not so useful so I ran in gdb and I had this information:

```sh
[Thread 0x7fffe5d296c0 (LWP 9315) exited]
[New process 9082]
[Inferior 1 (process 9082) exited with code 01]
(gdb) bt
No stack.
(gdb) 
quit
```

I am adding the [error.log](https://github.com/netdata/netdata/files/9556813/error.log) to help. I tried to run `netdata` two times with a `parent-child` setup to get more information.

ktsaou commented:
Ready for review. Please test this.

thiagoftsm commented:
Hello guys,

Compilation on `13.1-RELEASE-p2` looks fine and netdata runs on this environment, but streaming is still a blocker for some tests due [this](https://github.com/netdata/netdata/pull/13646#issuecomment-1245314021) issue .

ktsaou commented:
rebased to master

thiagoftsm commented:
Hello @ktsaou ,

I observed that alerts is having a different behavior with this PR. I used this [alert](https://raw.githubusercontent.com/netdata/netdata/master/tests/template_dimension/system_cpu.conf.alarm_foreach) to test and I got these different results:

![Screenshot_20220913_214346](https://user-images.githubusercontent.com/49162938/190015555-55863f5e-9f43-4a58-824a-5e99afc1738d.png)

As you can see using our current `master` branch I have a big sample of alerts, while using this PR during 15 minutes I only had 3 alerts. Is this really expected? Or do you have a bug?

Best regards!

thiagoftsm commented:
Hello @ktsaou  and @stelfrag ,

Today I ran some tests with this PR. I observed that independent of the memory mode the PR is running and it is possible:

- To access dashboard.
- See information on cloud.
- Use API V1 with our [script](https://github.com/netdata/netdata/tree/master/tests/urls)
- Use exporter (JSON).

, but issues with `streaming` and `health` that I reported are concerns.

ktsaou commented:
Guys, please don't ask me if there is a bug. Probably there is.
When you have something that you believe is not right, please try to find what is the difference.

For example, related to your observation @thiagoftsm , please pick any of the alarms the master shows and find out:

1. is it there at this PR?
2. is it calculated?
3. does the value seem right?

Then post your findings. If there is something that is wrong, please let me know of something more concrete that "hey, this seems wrong". Try to trace the issue...

ktsaou commented:
Tests that would be helpful:

1. I changed all dimension, chart and host variables (variables that can be used in alarms). The values should not be changed, but check indexing. Are the variables resolved? Do the alarms have a value? Does the value seem right?
2. I changed how alarms are created from templates. Are all alarms there?
3. I changed how alarm templates for-each-dimension is maintained. Can you test this kind of templates?
4. I changed how dimensions and charts are indexed. Are all the charts there? The dimensions?
5. I changed how charts obsoletion and archiving is triggered. Are charts being archieved?

Apart from all the above, any crash, any slowness, any problem with streaming, exporting, etc should be reported.

vlvkobal commented:
Valgrind reports a lot of leaks on Netdata exit so its output becomes very difficult to analyze. Are you planning to eliminate them in this PR?
```
==264798== LEAK SUMMARY:
==264798==    definitely lost: 267,984 bytes in 3,103 blocks
==264798==    indirectly lost: 5,250,581 bytes in 96,022 blocks
==264798==      possibly lost: 3,111,059 bytes in 42,261 blocks
```

They are mostly like
```
==264798== 539 bytes in 24 blocks are possibly lost in loss record 8,853 of 10,125
==264798==    at 0x4841888: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==264798==    by 0x14F9CD: mallocz (libnetdata.c:170)
==264798==    by 0x145B0C: string_index_insert (dictionary.c:1999)
==264798==    by 0x146ADD: string_strdupz (dictionary.c:2078)
==264798==    by 0x18B600: health_readfile (health_config.c:1072)
==264798==    by 0x15121C: recursive_config_double_dir_load (libnetdata.c:1451)
==264798==    by 0x18BCDF: health_readdir (health_config.c:1179)
==264798==    by 0x185B21: health_reload_host (health.c:177)
==264798==    by 0x185CAE: health_reload (health.c:206)
==264798==    by 0x1A938D: reload_host_labels (rrdhost.c:1349)
==264798==    by 0x1337AA: main (main.c:1492)
==264798==
==264798== 556 bytes in 10 blocks are definitely lost in loss record 8,858 of 10,125
==264798==    at 0x4841888: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==264798==    by 0x56F163E: strdup (strdup.c:42)
==264798==    by 0x14F93A: strdupz (libnetdata.c:159)
==264798==    by 0x1AC20B: rrdset_cache_dir (rrd.c:146)
==264798==    by 0x1B1B4F: rrdset_insert_callback (rrdset.c:151)
==264798==    by 0x146C43: item_create_unsafe (dictionary.c:817)
==264798==    by 0x14834C: dictionary_set_item_unsafe (dictionary.c:1154)
==264798==    by 0x148A91: dictionary_set_advanced (dictionary.c:1277)
==264798==    by 0x1ACA10: rrdset_index_add (rrdset.c:385)
==264798==    by 0x1AE370: rrdset_create_custom (rrdset.c:807)
==264798==    by 0x129664: workers_utilization_update_chart (global_statistics.c:1413)
==264798==    by 0x12A1FF: worker_utilization_charts (global_statistics.c:1858)
==264798==
==264798== 560 bytes in 16 blocks are possibly lost in loss record 8,861 of 10,125
==264798==    at 0x4841888: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==264798==    by 0x3457A5: JudyMalloc (JudyMalloc.c:43)
==264798==    by 0x344C19: insStrJudyLTree (JudyHS.c:352)
==264798==    by 0x345497: JudyHSIns (JudyHS.c:497)
==264798==    by 0x145AE7: string_index_insert (dictionary.c:1985)
==264798==    by 0x146ADD: string_strdupz (dictionary.c:2078)
==264798==    by 0x18938F: health_readfile (health_config.c:609)
==264798==    by 0x15121C: recursive_config_double_dir_load (libnetdata.c:1451)
==264798==    by 0x18BCDF: health_readdir (health_config.c:1179)
==264798==    by 0x1A5ECF: rrdhost_initialize_health (rrdhost.c:292)
==264798==    by 0x1A7C82: rrdhost_create (rrdhost.c:405)
==264798==    by 0x1A8DDC: rrd_init (rrdhost.c:894)
```

vlvkobal commented on collectors/proc.plugin/proc_net_dev.c:
```diff
@@ -1018,6 +1017,7 @@ int do_proc_net_dev(int update_every, usec_t dt) {
                             else rrdset_next(d->st_speed);
 
                             rrddim_set_by_pointer(d->st_speed, d->rd_speed, (collected_number)d->speed * KILOBITS_IN_A_MEGABIT);
+                            rrdsetvar_custom_chart_variable_set(d->st_bandwidth, d->chart_var_speed, (NETDATA_DOUBLE) d->speed * KILOBITS_IN_A_MEGABIT);
                             rrdset_done(d->st_speed);
                         }
```
```suggestion
                            rrdset_done(d->st_speed);
                        }

                        rrdsetvar_custom_chart_variable_set(d->st_bandwidth, d->chart_var_speed, (NETDATA_DOUBLE) d->speed * KILOBITS_IN_A_MEGABIT);
```

ktsaou commented:
Today, I rewrote the whole dictionary to support 2 independent locks for indexing and traversal. More than half of it changed...
The good think is that it is now cleaner and better to understand. Locks are used only when necessary.

Also, write locks in dictionary are now recursive. So all the unsafe versions of its API have been eliminated. Just do `dfe_start_write()` and then call `dictionary_set()` or `dictionary_del()`. No deadlock...

ktsaou commented:
@vlvkobal I fixed the leak in rrd.c (`RRDSET->cache_dir`).
The other 2 about health_config.c, I am not sure I see the leak. Valgrind says "possibly lost" for them. Can you help tracking them down?

ktsaou commented:
@thiagoftsm any news about the alarms discrepancies you had?

vlvkobal commented:
After the last changes, I've got
```
==410962== Thread 8 STREAM_RECEIVER:
==410962== Conditional jump or move depends on uninitialised value(s)
==410962==    at 0x1480F6: item_add_or_reset_value_and_acquire (dictionary.c:1087)
==410962==    by 0x1482EC: dictionary_set_and_acquire_item_advanced (dictionary.c:1331)
==410962==    by 0x1484D8: dictionary_set_advanced (dictionary.c:1345)
==410962==    by 0x1AA505: labels_add_already_sanitized (rrdlabels.c:552)
==410962==    by 0x1AA6A4: copy_label_to_dictionary_callback (rrdlabels.c:733)
==410962==    by 0x148FF1: dictionary_walkthrough_rw (dictionary.c:1663)
==410962==    by 0x1AB387: rrdlabels_migrate_to_these (rrdlabels.c:745)
==410962==    by 0x1AB6C5: rrdset_update_rrdlabels (rrdlabels.c:966)
==410962==    by 0x190EBA: pluginsd_clabel_commit_action (pluginsd_parser.c:195)
==410962==    by 0x192731: pluginsd_clabel_commit (pluginsd_parser.c:631)
==410962==    by 0x2902DB: parser_action (parser.c:295)
==410962==    by 0x280207: streaming_parser (receiver.c:419)
==410962==
==410962== Thread 12 PLUGIN[proc]:
==410962== Conditional jump or move depends on uninitialised value(s)
==410962==    at 0x1480F6: item_add_or_reset_value_and_acquire (dictionary.c:1087)
==410962==    by 0x1482EC: dictionary_set_and_acquire_item_advanced (dictionary.c:1331)
==410962==    by 0x19DF4C: rrdinstance_from_rrdset (rrdcontext.c:914)
==410962==    by 0x1A17C2: rrdcontext_updated_rrdset (rrdcontext.c:1404)
==410962==    by 0x1AC8B1: rrdset_react_callback (rrdset.c:355)
==410962==    by 0x1482AC: item_add_or_reset_value_and_acquire (dictionary.c:1089)
==410962==    by 0x1482EC: dictionary_set_and_acquire_item_advanced (dictionary.c:1331)
==410962==    by 0x1484D8: dictionary_set_advanced (dictionary.c:1345)
==410962==    by 0x1ACB8D: rrdset_index_add (rrdset.c:388)
==410962==    by 0x1AE4DC: rrdset_create_custom (rrdset.c:810)
==410962==    by 0x2D7567: do_proc_stat (proc_stat.c:1027)
==410962==    by 0x2A5663: proc_main (plugin_proc.c:153)
==410962==
==410962== Thread 15 PLUGINSD[ebpf]:
==410962== Conditional jump or move depends on uninitialised value(s)
==410962==    at 0x1480F6: item_add_or_reset_value_and_acquire (dictionary.c:1087)
==410962==    by 0x1482EC: dictionary_set_and_acquire_item_advanced (dictionary.c:1331)
==410962==    by 0x19DE75: rrdinstance_from_rrdset (rrdcontext.c:897)
==410962==    by 0x1A17C2: rrdcontext_updated_rrdset (rrdcontext.c:1404)
==410962==    by 0x1AC8B1: rrdset_react_callback (rrdset.c:355)
==410962==    by 0x1482AC: item_add_or_reset_value_and_acquire (dictionary.c:1089)
==410962==    by 0x1482EC: dictionary_set_and_acquire_item_advanced (dictionary.c:1331)
==410962==    by 0x1484D8: dictionary_set_advanced (dictionary.c:1345)
==410962==    by 0x1ACB8D: rrdset_index_add (rrdset.c:388)
==410962==    by 0x1AE4DC: rrdset_create_custom (rrdset.c:810)
==410962==    by 0x19098B: pluginsd_chart_action (pluginsd_parser.c:53)
==410962==    by 0x191A62: pluginsd_chart (pluginsd_parser.c:404)
==410962==
==410962== Conditional jump or move depends on uninitialised value(s)
==410962==    at 0x1480F6: item_add_or_reset_value_and_acquire (dictionary.c:1087)
==410962==    by 0x1482EC: dictionary_set_and_acquire_item_advanced (dictionary.c:1331)
==410962==    by 0x1484D8: dictionary_set_advanced (dictionary.c:1345)
==410962==    by 0x1ACB8D: rrdset_index_add (rrdset.c:388)
==410962==    by 0x1AE4DC: rrdset_create_custom (rrdset.c:810)
==410962==    by 0x19098B: pluginsd_chart_action (pluginsd_parser.c:53)
==410962==    by 0x191A62: pluginsd_chart (pluginsd_parser.c:404)
==410962==    by 0x2902DB: parser_action (parser.c:295)
==410962==    by 0x192EC8: pluginsd_process (pluginsd_parser.c:802)
==410962==    by 0x18FD06: pluginsd_worker_thread (plugins_d.c:248)
==410962==    by 0x15D7D7: thread_start (threads.c:196)
==410962==    by 0x56DB74C: start_thread (pthread_create.c:442)
```

ktsaou commented:
Rebased to master.

@vlvkobal addressed your findings.

ktsaou commented:
@vlvkobal could you please fix cmocka?

ktsaou commented:
rebased again

thiagoftsm commented:
Hello @ktsaou,

Sorry for the delay to answer you about [these questions](https://github.com/netdata/netdata/pull/13646#issuecomment-1247332884), but I had to focus in a collector I am working, else I could not advance with it. Now

1. I will retest everything relate to `alerts` to verify if we should have or not the alerts like it always happened with current master code.
2. Related to [streaming](https://github.com/netdata/netdata/pull/13646#issuecomment-1243021717), I discussed with @stelfrag, and I observed that issues with streaming are happening when I compile netdata with `libasan`, when I compile without it I do not have any issues. A little bit of additional details, I used `vacuum`, I removed the whole database and the issue persists. I detected the issue with this PR, because you rebase it.

thiagoftsm commented on database/rrdhost.c:
```diff
@@ -206,6 +200,115 @@ void set_host_properties(RRDHOST *host, int update_every, RRD_MEMORY_MODE memory
+    rrdcalc_rrdhost_index_init(host);
+    host->rrdvars = rrdvariables_create();
+
+    host->health_default_warn_repeat_every = config_get_duration(CONFIG_SECTION_HEALTH, "default repeat warning", "never");
```
Hello @ktsaou,

I am running different tests on your PR, and I coverity is reporting:

```sh
*** CID 357071:  High impact quality  (Y2K38_SAFETY)
/database/rrdhost.c: 254 in rrdhost_initialize_health()
248         host->health_default_warn_repeat_every = config_get_duration(CONFIG_SECTION_HEALTH, "default repeat warning", "never");
249         host->health_default_crit_repeat_every = config_get_duration(CONFIG_SECTION_HEALTH, "default repeat critical", "never");
250     
251         host->health_log.next_log_id = 1;
252         host->health_log.next_alarm_id = 1;
253         host->health_log.max = 1000;
>>>     CID 357071:  High impact quality  (Y2K38_SAFETY)
>>>     A "time_t" value is stored in an integer with too few bits to accommodate it.  The expression "now_realtime_sec()" is cast to "uint32_t".
254         host->health_log.next_log_id = (uint32_t)now_realtime_sec();
255         host->health_log.next_alarm_id = 0;
256     
257         long n = config_get_number(CONFIG_SECTION_HEALTH, "in memory max health log entries", host->health_log.max);
258         if(n < 10) {
259             error("Host '%s': health configuration has invalid max log entries %ld. Using default %u", rrdhost_hostname(host), n, host->health_log.max);

** CID 329593:    (TAINTED_SCALAR)
```

ktsaou replied:
Yes, this is by design.

thiagoftsm commented on database/rrdcalc.c:
```diff
@@ -252,139 +315,176 @@ inline void rrdsetcalc_unlink(RRDCALC *rc) {
+        rc->chart = string_dup(st->id);
+        uuid_copy(rc->config_hash_id, rt->config_hash_id);
+
+        rc->dimensions = (ctr->overwrite_dimensions) ? string_strdupz(ctr->overwrite_dimensions) : string_dup(rt->dimensions);
```
We also have this suggestion:

```sh
*** CID 357074:  Concurrent data access violations  (MISSING_LOCK)
/database/rrdcalc.c: 435 in rrdcalc_rrdhost_insert_callback()
429             rc->dimensions = (ctr->overwrite_dimensions) ? string_strdupz(ctr->overwrite_dimensions) : string_dup(rt->dimensions);
430             rc->foreach_dimension = NULL;
431             rc->foreach_dimension_pattern = NULL;
432     
433             rc->green = rt->green;
434             rc->red = rt->red;
>>>     CID 357074:  Concurrent data access violations  (MISSING_LOCK)
>>>     Accessing "rc->value" without holding lock "netdata_rwlock_t.rwlock_t". Elsewhere, "rrdcalc.value" is accessed with "netdata_rwlock_t.rwlock_t" held 5 out of 6 times.
435             rc->value = NAN;
436             rc->old_value = NAN;
437     
438             rc->delay_up_duration = rt->delay_up_duration;
439             rc->delay_down_duration = rt->delay_down_duration;
440             rc->delay_max_duration = rt->delay_max_duration;

** CID 357073:    (RESOURCE_LEAK)
/libnetdata/dictionary/dictionary.c: 1123 in item_add_or_reset_value_and_acquire()
/libnetdata/dictionary/dictionary.c: 1160 in item_add_or_reset_value_and_acquire()
```

ktsaou replied:
The "Concurrent data access violations" found is a false positive. `rrdcalc_rrdhost_insert_callback()` is the constructor of the object. No other thread has access to it.

The `item_add_or_reset_value_and_acquire()` are accessing the value of a different object. Probably coverity got confused.

thiagoftsm commented on libnetdata/dictionary/dictionary.c:
```diff
@@ -533,636 +751,816 @@ static inline int hashtable_delete_unsafe(DICTIONARY *dict, const char *name, si
+            // DONT DO IT THE OTHER WAY AROUND - DO NOT CROSS THE LOCKS!
+            dictionary_index_lock_unlock(dict);
+
+            item_linked_list_add(dict, item);
```
And this is the last for this PR:

```sh
*** CID 357073:    (RESOURCE_LEAK)
/libnetdata/dictionary/dictionary.c: 1123 in item_add_or_reset_value_and_acquire()
1117                 item_linked_list_add(dict, item);
1118                 added_or_updated = true;
1119             }
1120             else {
1121                 if (!item_check_and_acquire(dict, *item_pptr)) {
1122                     spins++;
>>>     CID 357073:    (RESOURCE_LEAK)
>>>     Variable "item_pptr" going out of scope leaks the storage it points to.
1123                     continue;
1124                 }
1125     
1126                 // the item is already in the index
1127                 // so, either we will return the old one
1128                 // or overwrite the value, depending on dictionary flags
/libnetdata/dictionary/dictionary.c: 1160 in item_add_or_reset_value_and_acquire()
1154                     // we did really nothing!
1155                     ;
1156                 }
1157     
1158                 dictionary_index_lock_unlock(dict);
1159             }
>>>     CID 357073:    (RESOURCE_LEAK)
>>>     Variable "item_pptr" going out of scope leaks the storage it points to.
1160         } while(!item);
1161     
1162     
1163         if(unlikely(spins > 0 && dict->stats))
1164             DICTIONARY_STATS_INSERT_SPINS_PLUS(dict, spins);
```

ktsaou replied:
This is how Judy works. It gives you a pointer to its own storage which you can modify its contents to put whatever you like.

thiagoftsm commented:
@stelfrag and @Ferroin I reported some issues related to this PR, but I also saw others that are not related, probably we are not running `covertiy` as we should.

thiagoftsm commented:
I will complement @vlvkobal report given my output for [valgrind](https://github.com/netdata/netdata/files/9590854/myfile.txt). We have issues related to this PR, but we also have ancient issues that began to be shown only after Valgrind `3.19` to be released.

thiagoftsm commented:
And now positive news, this PR ran as expected on:

| Machine | Distribution |
|-------------|------------------|
| X86_64 | Slackware Current|
| X86_64 VM | Arch Linux |
| X86_64 VM | FreeBSD 13.2 |
| X86_64 VM | Ubuntu 18..04 |
| X86_64 VM | CentOS 7.9 |
| Armvl7 | Raspberry Pi OS |

I tested the following features:

- Cloud 
- Exporter (JSON, Prometheus)
- Streaming
- Normal alerts were raised (I tested them with exporter)

I used `[db].mode` as `ram` and `dbengine`.
I will proceed with tests in the next days. I wanna see normal alerts while I am compiling kernel in few hours, and of course I will test "strange alerts"

ktsaou commented:
I had a bug and the destructors were not running in some cases. It is fixed now, so there should be no more memory leaks.

thiagoftsm commented:
As I said yesterday I compiled the kernel today and I had few alerts that were expected. I am compiling last commit now to verify with `valgrind` possible memory leaks.

thiagoftsm commented:
Hello @ktsaou,

I ran this PR in `valgrind` again, this time I used the following environment:

- I used valgrind only with parent.
- I used a child compiled on FreeBSD
- I also used our [script](https://github.com/netdata/netdata/tree/master/tests/urls) to do thousands of calls for our API.

I could see less leak reports from valgrind, but we still have few of them reported for `dictionary` library, please, take a look in this
[report](https://github.com/netdata/netdata/files/9592810/myfile.txt).

Best regards!

thiagoftsm commented:
Hello @ktsaou ,

Your PR is reporting conflict now, please, rebase it.

thiagoftsm commented:
Now about the issue with alerts, after to rebase your PR with last commits, I observed that including the alerts present in [this](https://github.com/netdata/netdata/tree/master/tests/template_dimension link) are also sent for users:

| Branch             | Data from API | Error log | Telegram notification |
|---------------------|---------------------|-------------|-----------------------------|
|Master              | [data_from_master.txt](https://github.com/netdata/netdata/files/9594873/data_from_master.txt) | [error_master.txt](https://github.com/netdata/netdata/files/9594875/error_master.txt)  | [telegram_master_notification.txt](https://github.com/netdata/netdata/files/9594876/telegram_master_notification.txt) |
| This PR           | [data_from_pr.txt](https://github.com/netdata/netdata/files/9594877/data_from_pr.txt) | [error_pr.txt](https://github.com/netdata/netdata/files/9594878/error_pr.txt) | [telegram_pr_notification.txt](https://github.com/netdata/netdata/files/9594879/telegram_pr_notification.txt) |

To generate this data I inserted the following debug message:

```c
error("KILLME %s: FLAG %u, VALUE %g, status %u", string2str(rc->name), rc->run_flags & RRDCALC_FLAG_DB_ERROR, rc->value, status);
```
in both branches. For master I insered at line https://github.com/netdata/netdata/blob/28f55f66147784233f14f1c0608e460b460d99c5/health/health.c#L1097 and for current PR at analogous position, because this PR is removing some lines

I am already running additional tests with this PR.

ktsaou commented:
merged the latest changes from master

ktsaou commented:
So, @thiagoftsm you see a problem or this PR is ok?

thiagoftsm commented:
> So, @thiagoftsm you see a problem or this PR is ok?

It was OK with the last round of tests, but I saw that you pushed more commits since my last round of tests, give me few hours to retest everything again, and as soon I finish I will approve the PR if there is not any issue. :handshake:

thiagoftsm commented:
Hello @ktsaou ,

I retested everything and the PR is working as expected.

I saw that we still have blank checkpoints, so I am waiting new commits to review and test.

ktsaou commented:
I think we can merge this and continue later.
Every merge on master now requires a lot of work on this PR.

So, let's merge it as it is.

thiagoftsm commented:
> I think we can merge this and continue later. Every merge on master now requires a lot of work on this PR.
> 
> So, let's merge it as it is.

Ok, I let the PR running on `Raspberry PI` during the whole night and I did not get any coredump. I will review the last commits and verify that nothing is broken.

MrZammler commented on libnetdata/log/log.c:
```diff
@@ -897,6 +905,10 @@ void fatal_int( const char *file, const char *function, const unsigned long line
+#ifdef NETDATA_INTERNAL_CHECKS
+    crash_netdata();
+#endif
+
```
Hi, is this staying after the merge? A way to get more info via the core dump when it fatals? (Just asking in case it was used for testing or similar).

ktsaou replied:
I suggest to keep it. It only runs on `fatal()` when running with `NETDATA_INTERNAL_CHECKS`.

ktsaou replied:
It allows us to get a core dump to examine the situation of the fatal. Otherwise, it is not easy to understand how the fatal was triggered.

thiagoftsm reviewed (approved):
Considering that we are not going to have more commits. I am approving this PR that is running as expected since last commit. During this period of time I tested `alerts`, `API`, `exporter`, `health`, and `cloud`. I used three different hosts (X86_64, Armvl7, X86_64 VM), and three different Linux distribution (Slackware Current, Raspberry PI OS, CentOS 7.9). LGTM!

vkalintiris commented on libnetdata/dictionary/dictionary.c:
@ktsaou @stelfrag  I recall us discussing the performance of this piece of code and I found this interesting https://youtu.be/rmGJc9PXpuE?t=16
